### PR TITLE
Fix arcsinh() returning -Infinity for large negative inputs

### PR DESCRIPTION
**Pull request by MaxFreedomPollard**
## Summary

`jax.numpy.arcsinh()` is wrong across the negative half of its domain. On cpu it gives `-Infinity` from `-1e8` down, `NaN` at `-Infinity` and `+0` at `-0`. On wasm, real float32, `arcsinh(-4096)` is already `-Infinity`, and it is visibly wrong earlier: `arcsinh(-1000)` gives `-7.624619007110596` against numpy float32 `-7.600902557373047`. The new gradient test fails on main too, where `grad(arcsinh)(-1e6)` is `NaN`.

```js
np.arcsinh(np.array([-1e8, -1e7, -Infinity])).js();
// actual (cpu):        [-Infinity, -16.805431, NaN]
// numpy 2.4.6 float32: [-19.113829, -16.811243, -Infinity]
```

The Array API spec requires `-0 -> -0`, `-infinity -> -infinity` and `NaN -> NaN` (https://data-apis.org/array-api/latest/API_specification/generated/array_api.asinh.html), and FEATURES.md marks arcsinh as supported.

## Cause

src/library/numpy.ts:3257 read `log(x.ref.add(sqrt(square(x).add(1))))`. For negative `x` that adds two nearly equal numbers of opposite sign, so once `sqrt(x^2 + 1)` rounds to `|x|`, at 2^12 in float32 and 2^26 in float64, the sum is zero and the logarithm is `-Infinity`.

## Changes

asinh is odd, so the new body takes the sign with `signbit()`, evaluates at the magnitude and multiplies it back, using a factor rather than `absolute()` so zero stays differentiable. Past `1e10` the `+1` is below the last bit of `x^2`, so that branch takes the `ln(2x)` limit, avoiding the float32 overflow near `1.8e19`. Each branch runs on an argument the other one handles, so an unselected infinity never reaches the gradient.

float64 on cpu is repaired too, where `arcsinh(-1e300)` returned `+Infinity`. Integer and bool input now promotes to float32 instead of throwing "Unsupported dtype for Sqrt", following JAX's x32 default and the promotions `expm1()` and `log1p()` already do here.

## Verification

Node 26, cpu and wasm only. This adds five arcsinh tests to test/numpy.test.ts and one to test/dtype-f64.test.ts, where there were none: 11 failed and 1237 passed against main's numpy.ts, 1248 passed here. Across 26 values from -1e25 to 1e25, cpu matches numpy 2.4.6 float32 exactly; wasm agrees to 1.192e-6 relative, worst at 1000 and -1000, where main returns the same `7.600893497467041` at `+1000`, so that gap is pre-existing. `pnpm test run --no-browser --exclude 'packages/**'` gives 30 files and 2050 passed; with packages/* included the only failure is packages/loaders/src/tokenizers.test.ts, which needs OPFS and fails the same way on main. `pnpm check`, `pnpm lint --max-warnings 0` and `pnpm format:check` are clean.

### CODE REVIEW

**Comment by ekzhang** [comment]
Hi, failing at extreme values is OK for this current implementation. If there's a need for stricter numerics for these functions then they should be implemented as primitives so they can remain performant rather than this impl which is a lot slower.
